Update dependencies

theopak opened the pull request:
I'm submitting this PR so unit tests run. There's a known security vulnerability in express 3.0.0 which is fixed in 4.0.0, according to david-dm. I think that should be updated even though express is only used in the test suite for this package.

I also updated non-dev dependencies in package.json to reference the most recent minor.patch in the `^Major.minor.patch` specifier, which has no functional difference because that's already what npm pulls due to use of `^`.
- Update devDependencies, including fixes to the testing setup's known security vulnerabilities. More info: https://david-dm.org/Cimpress-MCP/oauth_reverse_proxy#info=devDependencies&view=table
- Update (run) dependencies. More info: https://david-dm.org/Cimpress-MCP/oauth_reverse_proxy#info=dependencies&view=table

theopak committed:
Update devDependencies, including fixes to the testing setup's known security vulnerabilities. More info: https://david-dm.org/Cimpress-MCP/oauth_reverse_proxy#info=devDependencies&view=table

theopak committed:
Update (run) dependencies. More info: https://david-dm.org/Cimpress-MCP/oauth_reverse_proxy#info=dependencies&view=table

theopak committed:
Also update express.

ryanbreen commented:
Hmm, looks like npm test failed but called itself a success:

[2015-09-14T18:57:29.508Z] ERROR: oauth_reverse_proxy/2643 on testing-worker-linux-44a99d76-2-13018-linux-12-80297745: app.use() requires middleware functions (module=lib/index.js)
    TypeError: app.use() requires middleware functions
        at EventEmitter.use (/home/travis/build/Cimpress-MCP/oauth_reverse_proxy/node_modules/express/lib/application.js:209:11)
        at Object.<anonymous> (/home/travis/build/Cimpre

theopak commented:
Yup, weird that it counted as a pass. This isn't good to merge as-is.

theopak committed:
Require recent Node.js version for Travis-CI.

theopak committed:
Fix improper use of body-parser: https://github.com/expressjs/body-parser/blob/master/HISTORY.md#140--2014-06-19

theopak committed:
Update use of multer

theopak commented:
No idea why is reports 'passed' to github. 

https://travis-ci.org/Cimpress-MCP/oauth_reverse_proxy/builds/80303096

>   93 passing (30s)
>   276 failing

theopak committed:
Change Travis-CI test script

theopak committed:
Change Travis-CI install and test settings.

theopak commented:
The previous commit fixes #10.

theopak committed:
Fix issue where Travis-CI reported false successes because the voeralls command returns zero even when the test command before it returns non-zero.

theopak committed:
Alright I'm fixing the test reporting process for real this time by straight up removing bunyan. @ryanbreen why was it there in the first place?

theopak committed:
Put bunyan back in, but return PIPESTATUS[0]

theopak committed:
Remove reference to XML file format, which is not allowed for config files.

theopak committed:
Fix previously unnoticed failing unit test for proxy_manager.js

theopak committed:
Address use of up-to-date multer

theopak committed:
Fix outdated proxy_config test which resulted in this failure:

    2) bootstrapping should reject invalid proxy config file 1999_called_service.xml:
          AssertionError: expected 'undefined' to be 'string'
          + expected - actual
          -undefined
          +string

          at Assertion.fail (node_modules/should/lib/assertion.js:180:17)
          at Assertion.prop.value (node_modules/should/lib/assertion.js:65:17)
          at Context.<anonymous> (test/bootstrap_test.js:117:27)

theopak committed:
Resolve relative proxy keystore paths to absolute paths on the filesystem

theopak committed:
Multer

ryanbreen commented:
Still failing.  So did one of these minor number version bumps in node_modules bust us?

theopak commented:
Yes, I'm having trouble fixing one of the upgrades. I just commented on the line. 

Also, I just noticed some failure cases related to PR #9 which I'm working on fixing.

theopak committed:
Formatting

theopak committed:
Resolve keystone paths during Proxy() initial configuration, and add a validation check after the existing checks. TODO (@theopak): add explicit unit test for validating oauth_secret_dir existence and readability.

theopak committed:
fixup! Resolve keystone paths during Proxy() initial configuration, and add a validation check after the existing checks. TODO (@theopak): add explicit unit test for validating oauth_secret_dir existence and readability.

theopak committed:
typos

theopak committed:
Typos

theopak committed:
Fix deprecated express function: 'express deprecated res.sendfile: Use res.sendFile instead test/job_server/index.js:108:11'

theopak committed:
Fix failing test which returned:

    1) basic config validation should reject an attempt to init oauth_reverse_proxy with a config_dir referencing a nonexistent directory:
       expected Function { name: '' } to throw exception with a message matching 'ENOENT, no such file or directory \'./does-not-exist\'', but got 'ENOENT, no such file or directory '/Users/tpak/git/oauth_reverse_proxy/does-not-exist''
    AssertionError: expected Function { name: '' } to throw exception with a message matching 'ENOENT, no such file or directory \'./does-not-exist\'', but got 'ENOENT, no such file or directory 'does-not-exist''
        at Assertion.fail (node_modules/should/lib/assertion.js:180:17)
        at Assertion.prop.value (node_modules/should/lib/assertion.js:65:17)
        at Context.<anonymous> (test/abnormal_config_test.js:44:17)
        at Context.slate_cleaner (test/abnormal_config_test.js:22:5)

theopak committed:
Always resolve to absolute paths

theopak committed:
Resolve absolute paths for SSL certs as well

theopak committed:
Now: "104 passing (2s), 18 failing"

theopak committed:
Now: 305 passing (7s), 38 failing locally

theopak commented:
The latest travis build shows "328 passing (10s)," "32 failing." It seems like there are at least 2 sources of these errors:
- the upgrade [http-proxy](https://github.com/nodejitsu/node-http-proxy) 1.**1**.2 -> 1.**11**.2 in `oauth_reverse_proxy/lib`, possibly related to  https://github.com/nodejitsu/node-http-proxy/issues/867
- the upgrade [multer](https://github.com/expressjs/multer) 0.1.0 -> 1.0.3 in the testing use

theopak committed:
Bump to v1.5.10 and add npm details

theopak committed:
Satisfy rimraf version compatibility required by grunt

    npm install
    npm WARN unmet dependency /Users/tpak/git/oauth_reverse_proxy/node_modules/grunt requires rimraf@'~2.2.8' but will load
    npm WARN unmet dependency /Users/tpak/git/oauth_reverse_proxy/node_modules/rimraf,
    npm WARN unmet dependency which is version 2.4.3

theopak committed:
Minor

theopak committed:
Upgrade connect 2.30.2 -> 3.4.0 (stable), which included API-breaking changes such as removal of all connect() middleware, deprecation of connect.createServer(), and a change to the connect() function signature.

theopak committed:
Add connect-query middleware

theopak committed:
Fix middleware order

theopak committed:
Fix connect-query by removing it.

theopak committed:
Actually fix things for real this time 😑

theopak committed:
Cmon Travis you can do it

theopak committed:
Touch

theopak committed:
redundancy

theopak commented:
@ryanbreen I'm having a lot of trouble with this. Can you take a look? Most of the failing tests seems to indicate that the pipeline isn't serving requests correctly, or attaching certain middleware at all.

theopak commented on lib/proxy/index.js:
```diff
@@ -25,6 +29,144 @@ HTTPS_AGENT.maxSockets = 100;
+ * using the passed in parameters. `Object.(forward|reverse)_proxy === true` iff the middleware should be used for a
+ * forward or reverse proxy, respectively. Order matters!
+ */
+function getMiddlewareList(this_obj, is_reverse_proxy) {
```
@ryanbreen In a major update, connect() removed all middleware that it owned. It's expected that developers will use other packages to supply middleware objects to apply via `app.use()`. That's why I wrote this function to generate a list that produces a list of objects to use.

ryanbreen commented:
Cool, I should be able to play around with it this afternoon.

Commit:
Remove npm shrinkwrap and expose proxies from proxy_manager by file name, not by path.

Commit:
Fix a missing mutator in the forward proxy case and remove accidental next call from the final forward proxy middleware.

Commit:
Fix errors building the correct path on forward proxy requests.

Commit:
connect-restreamer is broken.  Inline a working version.

Commit:
Set keep-alive on outgoing SSL requests since it is now disabled by default.

Commit:
Make sure validate_target_cert setting gets passed to outbound proxy.

Commit:
Update multipart to match new multer syntax.

Commit:
The json parser should not be used since only form encoded content is signed by OAuth.

Commit:
Switch to a wrapper urlencoded bodyparser to avoid failures due to content type mismatch.

Commit:
Fix log level.

theopak committed:
Merge pull request #1 from ryanbreen/update-depdencies

Partial fixes

ryanbreen commented:
```
  368 passing (41s)
sh: 1: Bad substitution
npm ERR! Test failed.  See above for more details.
```

theopak committed:
testing build with different command

theopak commented:
@ryanbreen was this PR merged by mistake? EDIT: everything basically LGTM except the package.json value `"version": "1.5.10-2"` which I did not yet remove `-2` from.

ryanbreen commented:
No.  All tests pass, so it's way better than what was HEAD.

theopak commented:
:+1:

theopak commented on lib/proxy/mutators/urlencoded_bodyparser.js:
```diff
@@ -0,0 +1,19 @@
+module.exports = function(proxy) {
+  return function(req, res, next) {
+    body_parser(req, res, function(err) {
+      if (err && err.status === 413) return next(err);
```
Creates issue: `body_parser.urlencoded()` middleware throws an error response for content that is not charset === `utf-8` (with the dash), ever since our upgrade (we went body-parser 1.0.2->1.14.0) in PR #12.
